noir/scripts/sync.sh

```diff
@@ -0,0 +1,387 @@
+  # checkout to re-apply the fixup and the patches.
+  if [ $have == true ] && has_fixup_and_patch; then
+    return 1 # false
+  else
+    return 0 # true
+  fi
+}
+
+# Find the latest nightly tag in the upstream repo
+function latest_nightly {
+  git ls-remote --tags --sort -refname $NOIR_REPO_URL nightly-* | head -n 1 | awk '{split($2,a,"/"); print a[3]}'
+}
+
+# Create an empty marker commit to show that patches have been applied or put in a patch file.
+function commit_patch_marker {
+  git -C noir-repo commit -m "$PATCH_COMMIT_MSG" --allow-empty
+}
+
+# Apply the fixup script and any local patch file.
+function patch_repo {
+  log Applying fixup on noir-repo
+  # Redirect the `bb` reference to the local one.
+  scripts/sync-in-fixup.sh
+  git -C noir-repo add . && git -C noir-repo commit -m "$FIXUP_COMMIT_MSG" --allow-empty
+  # Apply any patch file.
+  if [ -f $NOIR_REPO_PATCH ]; then
+    log Applying patches from $NOIR_REPO_PATCH
+    git -C noir-repo am ../$NOIR_REPO_PATCH
+  else
+    log "No patch file to apply"
+  fi
+  # Create an empty marker commit to show that patches have been applied.
+  commit_patch_marker
+}
+
+# Check out a tag, branch or commit.
+function switch_repo {
+  ref=$1
+  log Switching noir-repo to $ref
+  git -C noir-repo fetch --tags --depth 1 origin
+  # If we try to switch to some random commit after a branch it might not find it locally.
+  git -C noir-repo fetch --depth 1 origin $ref || true
+  # Try to check out an existing branch, or remote commit.
+  if git -C noir-repo checkout $ref; then
+    # If it's a branch we just need to pull the latest changes.
+    if is_on_branch; then
+      git -C noir-repo pull --rebase
+    fi
+  else
+    # If the checkout failed, then it should be a remote branch or tag
+    git -C noir-repo checkout --track origin/$ref
+  fi
+  # If we haven't applied the patch yet, we have to do it (again).
+  if ! has_patch_commit; then
+    patch_repo
+  else
+    log "Patches already applied"
+  fi
+}
+
+# Clone the repository if it doesn't exist.
+function init_repo {
+  url=$NOIR_REPO_URL
+  ref=$(read_wanted_ref)
+  if [ -d noir-repo ] && [ ! -d noir-repo/.git ]; then
+    # In all probability we just have some build leftovers after switching branches on aztec-packages,
+    # but play it safe and preserve them: instead of deleting `noir-repo` just re-initialize it.
+    git -C noir-repo init
+    git -C noir-repo remote add origin $url
+    switch_repo $ref
+  elif [ ! -d noir-repo ]; then
+    log Initializing noir-repo to $ref
+    # If we're cloning from a tag with --depth=1, we won't be able to switch to a branch later.
+    # On CI we won't be switching branches, but on dev machines we can, so there make a full checkout.
+    depth=$([ ! -z "${CI_FULL:-}" ] && echo "--depth 1" || echo "")
+    # `--branch` doesn't work for commit hashes
+    git clone $depth --branch $ref $url noir-repo \
+    || git clone $url noir-repo && git -C noir-repo checkout $ref
```

Comment:
```suggestion
    || git clone $url noir-repo && git -C noir-repo checkout -c $ref
```

This turns off the "detached HEAD" warning.

Reply:
Nice, thanks. I saw the long form in `git.rs` in Noir, but didn't want to increase the noise.